pkg/bpf/general/tc.go

```diff
@@ -0,0 +1,174 @@
+		syscall.S_IRUSR|syscall.S_IWUSR|syscall.S_IXUSR|
+			syscall.S_IRGRP|syscall.S_IXGRP); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	if err := os.MkdirAll(info.BpfFsPath,
+		syscall.S_IRUSR|syscall.S_IWUSR|syscall.S_IXUSR|
+			syscall.S_IRGRP|syscall.S_IXGRP); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (tc *BpfTCGeneral) loadKmeshTCObjects() (*ebpf.CollectionSpec, *ebpf.CollectionSpec, error) {
+	var (
+		errTcMarkDecrypt  error
+		errTcMarkEncrypt  error
+		specTcMarkDecrypt *ebpf.CollectionSpec
+		optsTcMarkDecrypt ebpf.CollectionOptions
+		specTcMarkEncrypt *ebpf.CollectionSpec
+		optsTcMarkEncrypt ebpf.CollectionOptions
+	)
+
+	optsTcMarkDecrypt.Maps.PinPath = tc.InfoTcMarkDecrypt.MapPath
+	optsTcMarkEncrypt.Maps.PinPath = tc.InfoTcMarkEncrypt.MapPath
+	if helper.KernelVersionLowerThan5_13() {
+		specTcMarkDecrypt, errTcMarkDecrypt = general.LoadKmeshTcMarkDecryptCompat()
+		specTcMarkEncrypt, errTcMarkEncrypt = general.LoadKmeshTcMarkEncryptCompat()
+	} else {
+		specTcMarkDecrypt, errTcMarkDecrypt = general.LoadKmeshTcMarkDecrypt()
+		specTcMarkEncrypt, errTcMarkEncrypt = general.LoadKmeshTcMarkEncrypt()
+	}
+	if errTcMarkDecrypt != nil {
+		return nil, nil, errTcMarkDecrypt
+	}
+	if errTcMarkEncrypt != nil {
+		return nil, nil, errTcMarkEncrypt
+	}
+	if specTcMarkDecrypt == nil || specTcMarkEncrypt == nil {
+		return nil, nil, fmt.Errorf("error: loadKmeshTCObjects() spec is nil")
+	}
+
+	utils.SetMapPinType(specTcMarkDecrypt, ebpf.PinByName)
+	if err := specTcMarkDecrypt.LoadAndAssign(&tc.KmeshTcMarkDecryptObjects, &optsTcMarkDecrypt); err != nil {
+		return nil, nil, err
+	}
+
+	utils.SetMapPinType(specTcMarkEncrypt, ebpf.PinByName)
+	if err := specTcMarkEncrypt.LoadAndAssign(&tc.KmeshTcMarkEncryptObjects, &optsTcMarkEncrypt); err != nil {
+		return nil, nil, err
+	}
+
+	return specTcMarkDecrypt, specTcMarkEncrypt, nil
+}
+
+func (tc *BpfTCGeneral) LoadTC() error {
+	if tc == nil {
+		return nil
+	}
+	specMarkDecrypt, specMarkEncrypt, err := tc.loadKmeshTCObjects()
+	if err != nil {
+		return err
+	}
+
+	prog := specMarkDecrypt.Programs[constants.TC_MARK_DECRYPT]
+	tc.InfoTcMarkDecrypt.Type = prog.Type
+	tc.InfoTcMarkDecrypt.AttachType = prog.AttachType
+
+	prog = specMarkEncrypt.Programs[constants.TC_MARK_ENCRYPT]
+	tc.InfoTcMarkEncrypt.Type = prog.Type
+	tc.InfoTcMarkEncrypt.AttachType = prog.AttachType
+
+	return nil
+}
+
+func (tc *BpfTCGeneral) Close() error {
+	if tc == nil {
+		return nil
+	}
+	if err := tc.KmeshTcMarkDecryptObjects.Close(); err != nil {
+		return err
+	}
+	progVal := reflect.ValueOf(tc.KmeshTcMarkDecryptObjects.KmeshTcMarkDecryptPrograms)
+	if err := utils.UnpinPrograms(&progVal); err != nil {
+		return err
+	}
+
+	mapVal := reflect.ValueOf(tc.KmeshTcMarkDecryptObjects.KmeshTcMarkDecryptMaps)
+	if err := utils.UnpinMaps(&mapVal); err != nil {
+		return err
+	}
+
+	if err := os.RemoveAll(tc.InfoTcMarkDecrypt.BpfFsPath); err != nil && !os.IsNotExist(err) {
```

Comment:
IIRC, BpfFsPath is same as below, so if you remove it here, you cannot unpin encrypt prog below, right?